Readme.md
# rs232com

## Documentation

## Installation


ChangeLog
ChangeLog
============

- v0.0.1

  - First time creation


tests/test_rs232com.py

import os
import sys

parent_path = os.path.join(os.path.dirname(__file__), '..')
sys.path.append(parent_path)
if True:
    import rs232com as com


if __name__ == '__main__':
    def on_open(ws, inst_com):
        print("### Open com port ###")

    def on_recv(ws, message):
        print(message)

    def on_error(ws, error):
        print(error)

    def on_close(ws):
        print("### closed ###")

    inst = com.rs232_com(
        on_open=on_open,
        on_recv=on_recv,
        on_error=on_error,
        on_close=on_close,
    )

    inst.open(
        port="COM4",
        baudrate=9600,
        bytesize=8,
        parity='None',
        stopbits=1,
    )

    inst.send(
        'COM4 open!!'
    )

    inst.recv()


tests/SerialComProcess.py
from queue import Empty
from multiprocessing import freeze_support, set_start_method
from multiprocessing import Process
from multiprocessing import Queue as MP_Queue

import tkinter as tk
from tkinter import ttk
import time

import rs232com as com  # nopep8

COM_PARAM = 'COM_PARAM'
RECV_DATA = 'RECV_DATA'
SEND_DATA = 'SEND_DATA'


class SerialComProcess():
    def __init__(self):
        self.inerval_time = 0.1
        self.instCom = None
        return

    def run(self, hQueue_recv, hQueue_send):
        self.queue_recv, self.queue_send = hQueue_recv, hQueue_send

        self.isLoop = True

        while(self.isLoop):
            self.get_Queue()
            time.sleep(self.inerval_time)

        return

    def get_Queue(self):
        while(not self.queue_recv.empty()):
            item = self.queue_recv.get()
            if COM_PARAM in item:
                param = item[COM_PARAM]

                com_port = param['com_port']
                boudrate = param['boudrate']
                bytesize = param['bytesize']
                stopbits = param['stopbits']
                parity = param['parity']

                self.instCom = com.rs232_com(
                    on_open=self.on_open,
                    on_recv=self.on_recv,
                    on_error=self.on_error,
                    on_close=self.on_close,
                )
                self.inst_comport = self.instCom.open(
                    port=com_port,
                    baudrate=boudrate,
                    bytesize=bytesize,
                    stopbits=stopbits,
                    parity=parity,
                )

            if SEND_DATA in item:
                send_data = item[SEND_DATA]
                if self.instCom:
                    self.instCom.send(send_data)

        if self.instCom:
            self.instCom.recv(is_rep=True)

        return

    def on_open(self, ws, inst_com):
        print("### Open com port ###")

    def on_recv(self, ws, message):
        self.queue_send.put(
            {RECV_DATA: message}
        )

    def on_error(self, ws, error):
        print(error)

    def on_close(self, ws):
        print("### closed ###")

    def join(self):
        self.isLoop = False


class SerialComTool():
    def __init__(self):
        self.isConnect = False
        self.interval_time = 200
        self.inst_comport = None

        self.queue_send, self.queue_recv = MP_Queue(), MP_Queue()

    def create_dialog(self):

        frame_conf = tk.LabelFrame(
            self.dialog,
            text='Serial Configuration'
        )

        frame_1 = tk.Frame(
            frame_conf
        )

        frame_2 = tk.Frame(
            frame_conf
        )

        frame_3 = tk.Frame(
            frame_conf
        )

        label_com = tk.Label(
            frame_1,
            text='COM Port:'
        )

        self.combo_com = ttk.Combobox(
            frame_1,
            values=self.com_ports,
            width=8,
            state='readonly',
        )

        label_boudrate = tk.Label(
            frame_2,
            text='baudrate:'
        )

        self.combo_boudrate = ttk.Combobox(
            frame_2,
            values=com.BAUDRATE_CHOICES,
            width=8,
            state='readonly',
        )
        self.combo_boudrate.current(12)

        label_bytesize = tk.Label(
            frame_2,
            text='bytesize:'
        )

        self.combo_bytesize = ttk.Combobox(
            frame_2,
            values=com.BYTESIZE_CHOICES,
            width=2,
            state='readonly',
        )
        self.combo_bytesize.current(3)

        label_parity = tk.Label(
            frame_2,
            text='parity:'
        )

        self.combo_parity = ttk.Combobox(
            frame_2,
            values=com.PARITY_CHOICES,
            width=6,
            state='readonly',
        )
        self.combo_parity.current(0)

        label_stopbits = tk.Label(
            frame_2,
            text='stopbits:'
        )

        self.combo_stopbits = ttk.Combobox(
            frame_2,
            values=com.STOPBITS_CHOICES,
            width=3,
            state='readonly',
        )
        self.combo_stopbits.current(0)

        self.btn_connect = tk.Button(
            frame_3,
            text='connect',
            width=10,
            command=self.btn_connect_clicked
        )

        frame_send = tk.LabelFrame(
            self.dialog,
            text='Send'
        )

        self.entry_send = tk.Entry(
            frame_send,
            width=89,
        )

        self.btn_send = tk.Button(
            frame_send,
            text='send',
            width=6,
            command=self.btn_send_clicked
        )

        frame_recv = tk.LabelFrame(
            self.dialog,
            text='Recv'
        )

        self.Text_recv = tk.Text(
            frame_recv,
            width=75, height=10
        )

        self.v_scroll = ttk.Scrollbar(
            frame_recv,
            orient=tk.VERTICAL,
            command=self.Text_recv.yview
        )
        self.Text_recv['yscrollcommand'] = self.v_scroll.set

        self.h_scroll = ttk.Scrollbar(
            frame_recv,
            orient=tk.HORIZONTAL,
            command=self.Text_recv.xview
        )
        self.Text_recv['xscrollcommand'] = self.h_scroll.set

        sp = 5
        frame_conf.pack(side=tk.TOP, fill=tk.X, padx=sp, pady=sp)
        frame_1.pack(side=tk.TOP, fill=tk.X, padx=sp, pady=sp)
        frame_2.pack(side=tk.TOP, fill=tk.X, padx=sp, pady=sp)
        frame_3.pack(side=tk.TOP, fill=tk.X, padx=sp, pady=sp)
        label_com.pack(side=tk.LEFT)
        self.combo_com.pack(side=tk.LEFT)
        label_boudrate.pack(side=tk.LEFT)
        self.combo_boudrate.pack(side=tk.LEFT)
        label_bytesize.pack(side=tk.LEFT)
        self.combo_bytesize.pack(side=tk.LEFT)
        label_parity.pack(side=tk.LEFT)
        self.combo_parity.pack(side=tk.LEFT)
        label_stopbits.pack(side=tk.LEFT)
        self.combo_stopbits.pack(side=tk.LEFT)
        self.btn_connect.pack(side=tk.LEFT)

        frame_send.pack(side=tk.TOP, fill=tk.X, padx=sp,
                        pady=sp, ipadx=sp, ipady=sp)
        self.entry_send.pack(side=tk.LEFT)
        self.btn_send.pack(side=tk.LEFT)

        frame_recv.pack(side=tk.TOP, fill=tk.X, padx=sp,
                        pady=sp, ipadx=sp, ipady=sp)
        self.entry_send.pack()
        self.v_scroll.pack(side=tk.RIGHT, fill=tk.Y)
        self.h_scroll.pack(side=tk.BOTTOM, fill=tk.X)
        self.Text_recv.pack(side=tk.LEFT, fill=tk.BOTH, expand=True)

        self.dialog.protocol("WM_DELETE_WINDOW", self.quit)

        return

    def after(self):
        while(not self.queue_recv.empty()):
            item = self.queue_recv.get()
            if RECV_DATA in item:
                message = item[RECV_DATA]
                self.Text_recv.insert(tk.END, message)

        self.after_id = self.dialog.after(self.interval_time, self.after)

    def btn_connect_clicked(self):
        if self.isConnect:
            self.isConnect = False
            self.btn_connect['text'] = 'connect'
            self.close()
        else:
            com_port = self.combo_com.get()
            if com_port in self.com_ports:
                self.open()
                self.isConnect = True
                self.btn_connect['text'] = 'disconnect'

        return

    def btn_send_clicked(self):
        send_data = self.entry_send.get()
        if self.isConnect:
            self.queue_send.put({SEND_DATA: send_data})
        return

    def open(self):
        com_port = self.combo_com.get()
        boudrate = int(self.combo_boudrate.get())
        bytesize = int(self.combo_bytesize.get())
        parity = self.combo_parity.get()
        stopbits = int(self.combo_stopbits.get())

        if com_port in self.com_ports:
            self.instProcess = SerialComProcess()
            self.instComProcess = Process(
                name='SerialComTool',
                target=self.instProcess.run,
                args=(
                    self.queue_send,
                    self.queue_recv,
                )
            )
            self.instComProcess.start()

            set_dict = {
                'com_port': com_port,
                'boudrate': boudrate,
                'bytesize': bytesize,
                'parity': parity,
                'stopbits': stopbits,
            }

            self.queue_send.put({COM_PARAM: set_dict})

            self.isConnect = True

    def close(self):
        # self.instComProcess.join()
        pass

    def quit(self):
        self.dialog.after_cancel(self.after_id)
        self.dialog.quit()

        if self.isConnect:
            self.close()

    def run(self):
        self.com_ports = com.get_port()

        self.dialog = tk.Tk()
        self.dialog.geometry('640x440')
        self.create_dialog()
        self.after_id = self.dialog.after(self.interval_time, self.after)
        self.dialog.mainloop()

        return


if __name__ == '__main__':
    freeze_support()
    set_start_method('spawn')

    instCom = SerialComTool().run()


rs232com/_exceptions.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

"""
FileName:
--------------------------------------------------------------------------------
    _exceptions.py

Description:
--------------------------------------------------------------------------------
    独自例外エラー定義

History:
--------------------------------------------------------------------------------
    2022/02/01 作成

"""


class RS232comException(Exception):
    pass


class RS232comPortException(RS232comException):
    pass


class RS232comParameterException(RS232comException):
    pass


class RS232comValueException(RS232comException):
    pass


class RS232comArgumentException(RS232comException):
    pass


class RS232comOpneException(RS232comException):
    pass


class RS232comCloseException(RS232comException):
    pass


class RS232comSendException(RS232comException):
    pass


rs232com/__main__.py
import time

import serial
from serial.tools import list_ports

from ._exceptions import *  # nopep8


BAUDRATE_CHOICES = [
    50, 75, 110, 134, 150, 200, 300, 600,
    1200, 1800, 2400, 4800, 9600,
    19200, 38400, 57600, 115200,
]

BYTESIZE_CHOICES = [
    serial.FIVEBITS,
    serial.SIXBITS,
    serial.SEVENBITS,
    serial.EIGHTBITS,
]

PARITY_CHOICES = [
    serial.PARITY_NAMES[serial.PARITY_NONE],
    serial.PARITY_NAMES[serial.PARITY_EVEN],
    serial.PARITY_NAMES[serial.PARITY_ODD],
]

STOPBITS_CHOICES = [
    serial.STOPBITS_ONE,
    serial.STOPBITS_ONE_POINT_FIVE,
    serial.STOPBITS_TWO,
]

XONXOFF_CHOICES = [
    serial.XON,
    serial.XOFF,
]


def get_port():
    """ シリアルComポートの取得

    Returns:
        [list]: シリアルComポートの文字列
    """

    ports = list_ports.comports()

    com_port = [info.device for info in ports]

    return com_port


class rs232_com():
    def __init__(self,
                 on_open=None, on_recv=None, on_error=None, on_close=None):

        self.is_connect = False
        self.inst_comport = None

        self.on_open = on_open
        self.on_recv = on_recv
        self.on_error = on_error
        self.on_close = on_close

        return

    def _callback(self, callback, *args):
        if callback:
            try:
                callback(self, *args)
            except Exception as e:
                if self.on_error:
                    self.on_error(self, e)

    def _chk_serial_parameter_(self, port: str, **kwargs):
        """ シリアルCOMポートの初期化

        Args:
            port (str): COMポート名
            **kwargs: 以下のkeywordパラメータを指定可能
                "baudrate": ビットレート,
                "bytesize": バイトサイズ(5,6,7,8),
                "parity": パリティ(None,Even,Odd),
                "stopbits": ストップビット指定(1.0/1.5/2.0),
                "xonxoff": (bool)Xon/Xoff制御,
                "rtscts": (bool)RTS/CTS制御,
                "dsrdtr": (bool)DSR/DTR制御,
                "timeout": (float)受信タイムアウト,

        Returns:
            [type]: [description]


        Raises:
            RS232comPortException: port指定の違反
            RS232comArgumentException: オプション指定引数の違反
            RS232comParameterException: オプション指定パラメータの範囲違反

        Returns:
            param(dict): 指定パラメータの辞書

        """
        param_dict = dict()

        parameters = {
            "baudrate": BAUDRATE_CHOICES,
            "bytesize": BYTESIZE_CHOICES,
            "parity": PARITY_CHOICES,
            "stopbits": STOPBITS_CHOICES,
            "xonxoff": [True, False],
            "rtscts": [True, False],
            "dsrdtr": [True, False],
        }

        com_ports = get_port()

        if port not in com_ports:
            raise RS232comPortException(
                "The specified Port is a violation."
            )

        args = list(parameters.keys()) + ['timeout']

        for key in kwargs:
            if key not in args:
                raise RS232comArgumentException(
                    f"{key}:Invalid argument"
                )
            else:
                if key in parameters:
                    param = kwargs[key]
                    param_range = parameters[key]
                    if param not in param_range:
                        raise RS232comParameterException(
                            f'{key}:Invalid parameter.'
                        )
                    else:
                        if key == "xonxoff":
                            if param:
                                param = XONXOFF_CHOICES[0]
                            else:
                                param = XONXOFF_CHOICES[1]

                        param_dict.setdefault(key, param)

                elif key == 'timeout':
                    timeout = kwargs['timeout']
                    if isinstance(timeout, float) or isinstance(timeout, int):
                        if timeout > 0:
                            param_dict.setdefault('timeout', timeout)
                        else:
                            raise RS232comParameterException(
                                f'{key}:Invalid parameter.'
                            )

        if 'timeout' not in param_dict:
            param_dict.setdefault('timeout', 0.3)

        return param_dict

    def open(self, port, **kwargs):
        """ シリアルComポートをOpen

        Raises:
            RS232comOpneException: シリアルポートOpenエラー

        Returns:
            [type]: [description]
        """

        self.port = port

        try:
            param = self._chk_serial_parameter_(port, **kwargs)
        except Exception as e:
            raise RS232comOpneException(f"Error:Serial open error {e}")

        try:
            inst_comport = serial.Serial(self.port)
        except Exception as e:
            raise RS232comOpneException(f"Error:Serial open error {e}")

        if isinstance(param, dict):
            for key, param in param.items():
                if key == "baudrate":
                    inst_comport.baudrate = param
                if key == "bytesize":
                    inst_comport.bytesize = param
                if key == "parity":
                    inst_comport.parity = param[0]
                if key == "stopbits":
                    inst_comport.stopbits = param
                if key == "xonxoff":
                    inst_comport.xonxoff = param
                if key == "rtscts":
                    inst_comport.rtscts = param
                if key == "dsrdtr":
                    inst_comport.dsrdtr = param
                if key == "timeout":
                    inst_comport.timeout = param

        self.is_connect = True
        self.inst_comport = inst_comport

        self._callback(self.on_open, inst_comport)

        return self.inst_comport

    def close(self):
        """ ComポートClose

        Raises:
            RS232comCloseException: シリアルポートCloseエラー

        Returns:
            [bool]: [description]
        """
        try:
            self.inst_comport.close()
        except Exception as e:
            raise RS232comCloseException(f"Error:Serial close error {e}")

        self.is_connect = False
        self.inst_comport = None

        self._callback(self.on_close)

        return True

    def send(self, send_data):
        """ 送信処理

        Args:
            send_data ([type]): [description]

        Raises:
            RS232comValueException: [description]
            RS232comSendException: [description]

        Returns:
            [type]: [description]

        """
        if self.is_connect:
            if isinstance(send_data, bytes):
                pass
            elif isinstance(send_data, str):
                try:
                    send_data = send_data.encode('utf-8')
                except Exception as e:
                    raise RS232comValueException(f"Error:{e}")
            elif isinstance(send_data, int) or isinstance(send_data, float):
                send_data = str(send_data)
                send_data = send_data.encode('utf-8')

            try:
                self.inst_comport.write(send_data)
            except Exception as e:
                raise RS232comSendException(f"Error:{e}")

        else:
            self._callback(
                self.on_error,
                RS232comSendException("Error:RS232C port is not open.")
            )

        return True

    def recv(self, is_rep=False):
        """ 受信処理

        Returns:
            [bytes]: 受信したデータ
        """

        while self.inst_comport.is_open:
            try:
                read_byte = self.inst_comport.read()
            except Exception as e:
                self._callback(self.on_error, e)
            if read_byte:
                self._callback(self.on_recv, read_byte)

            if is_rep:
                break

            time.sleep(0.05)

        return


if __name__ == '__main__':
    def on_open(ws, inst_com):
        print("### Open com port ###")

    def on_recv(ws, message):
        print(message)

    def on_error(ws, error):
        print(error)

    def on_close(ws):
        print("### closed ###")

    inst = rs232_com(
        on_open=on_open,
        on_recv=on_recv,
        on_error=on_error,
        on_close=on_close,
    )

    inst.open(
        port="COM4",
        baudrate=9600,
        bytesize=8,
        parity='None',
        stopbits=1,
    )

    inst.send(
        'COM4 open!!'
    )

    inst.recv()
